config: break dependencies on schema registry enums

dotnwat opened the pull request:
The `config::` sub-system is unique in that it is depended on by nearly all other sub-systems. As such, it becomes problematic when `config::` itself depends on sub-systems because cycles are created.

Generally these dependencies are only present in order to access sub-system specific types, such as enums.

In https://github.com/redpanda-data/redpanda/commit/0208f20d06114db3a19a6137eed710a366a9442f we fixed such a cycle between `config::` and `cloud_storage::` by mirror the shared type into `config::` and allowing `config::` to fully own this type. In the `cloud_storage::` sub-system, no changes are needed if the `config::` type is inspected (as is usually the case). Then when the value needs to cross sub-system boundaries, we introduced a `from_config(config::T)` pattern.

This is repeated here for a few schema registry types.


<!--
See https://github.com/redpanda-data/redpanda/blob/dev/CONTRIBUTING.md#pull-request-body
for more details and examples of what is expected in a PR body.

Content in this top section is REQUIRED. Describe, in plain language, the motivation
behind the change (bug fix, feature, improvement) in this PR and how the included
commits address it.

Add the GitHub keyword `Fixes` to link to bug(s) this PR will fix, e.g.
  Fixes #ISSUE-NUMBER, Fixes #ISSUE-NUMBER, ...

If this PR is a backport, link to the original with `Backport of PR`, e.g.
  Backport of PR #PR-NUMBER
-->

## Backports Required

<!-- Checking at least one of the checkboxes is REQUIRED if this PR is not a backport. -->

- [x] none - not a bug fix
- [ ] none - this is a backport
- [ ] none - issue does not exist in previous branches
- [ ] none - papercut/not impactful enough to backport
- [ ] v24.1.x
- [ ] v23.3.x
- [ ] v23.2.x

## Release Notes

<!--
If the changes in this PR do not need to be mentioned in the release
notes, then don't add a sub-section and simply list `none`, e.g.

Otherwise, adding a sub-section or `none` is REQUIRED if the PR is not a backport PR.
If this is a backport PR, adding contents to this section will override
the release notes section inherited from the original PR to dev.

Add one or more of the sub-sections with a short description bullet
point of the change, e.g.

### Bug Fixes

* Short description of the bug fix if this is a PR to `dev` branch.

### Features

* Short description of the feature. Explain how to configure.

### Improvements

* Short description of how this PR improves existing behavior.

-->

* none

dotnwat commented on src/v/config/convert.h:
```diff
@@ -475,37 +474,6 @@ struct convert<model::cloud_storage_chunk_eviction_strategy> {
 };
 
-template<>
-struct convert<pandaproxy::schema_registry::subject_name_strategy> {
```
@BenPope all indications are that this was dead code, and wasn't triggered by any sort of external input. still, i was surprised.

BenPope replied:
> @BenPope all indications are that this was dead code, and wasn't triggered by any sort of external input. still, i was surprised.

Good catch. At [some point during the implementation](https://github.com/redpanda-data/redpanda/pull/10650#issuecomment-1563757485) it was decided that it was unnecessary to be able to change the global default.

dotnwat commented on src/v/config/types.h:
```diff
@@ -51,4 +51,28 @@ inline std::ostream& operator<<(std::ostream& os, const s3_url_style& us) {
     }
 }
 
+enum class schema_id_validation_mode {
```
the `config::` system now fully owns this as part of its "external" interface with the outside work. the mirrored type in schema registry land still exists, but is fully internal and can have its own representation if it wants.

BenPope replied:
What are your thoughts on having
```
v_schema_registry_config
v_schema_registry
   DEPENDS v_schema_registry_config
v_config
   DEPENDS v_schema_registry_config
```
It's the [pattern used in v_security](https://github.com/redpanda-data/redpanda/blob/dev/src/v/security/CMakeLists.txt)

dotnwat replied:
i don't have any strong opinions about either direction. the only downside i see to the subsys_config verison is that one has to be extra disciplined in separating things within subsys to avoid accidentally rebuilding the entire tree since config:: will depend on all _config libraries.

i went the direction in this PR to make that separation very clear. but actually having the duplicated enum is only out of being practical / time constraint. i'm not sure why at this point the config option can't just be a string with a couple helpers to convert.

dotnwat commented on src/v/pandaproxy/schema_registry/validation.cc:
```diff
@@ -451,7 +451,8 @@ std::optional<schema_id_validator> maybe_make_schema_id_validator(
   const model::topic& topic,
   const cluster::topic_properties& props) {
-    auto mode = config::shard_local_cfg().enable_schema_id_validation();
+    auto mode = from_config(
+      config::shard_local_cfg().enable_schema_id_validation());
```
there are plenty of places in the tree that interrogate the config:: enable_schema_id_validation value, but this is the only place where config/external representation is converted into internal sub-system representation.

dotnwat commented on src/v/config/rjson_serialization.cc:
```diff
@@ -80,7 +80,7 @@ void rjson_serialize(
 
 void rjson_serialize(
-  json::Writer<json::StringBuffer>& w, const custom_aggregate& v) {
+  json::Writer<json::StringBuffer>& w, const test_custom_aggregate& v) {
```
i didn't intend to include this commit in this PR but it still needs to go in: "core" code should not be including anything from a test directory.

dotnwat commented on src/v/config/rjson_serialization.h:
```diff
@@ -15,14 +15,29 @@
 #include "config/data_directory_path.h"
 #include "config/endpoint_tls_config.h"
 #include "config/seed_server.h"
-#include "config/tests/custom_aggregate.h"
```
i didn't intend to include this commit in this PR but it still needs to go in: "core" code should not be including anything from a test directory.

vbotbuildovich commented:
<!--NEW FAILURES-->
new failures in https://buildkite.com/redpanda/redpanda/builds/49512#018fa8fe-cdc2-4764-adb5-cc3fea1f2b41:
```
"rptest.tests.cluster_config_test.ClusterConfigTest.test_rpk_edit_string"
```
<!--be08028b-f5f8-4e0b-803c-bb9d3ad70a7c-->

pgellert commented on src/v/config/rjson_serialization.h:
```diff
@@ -15,14 +15,29 @@
 
 #include <seastar/core/sstring.hh>
 
+/*
```
nit: is it worth wrapping this under a `testing` namespace?

dotnwat replied:
yeh good idea. fixed

michael-redpanda commented:
Looks like the changes broke `test_rpk_export_import`

oleiman reviewed (approved):
lgtm

dotnwat commented:
@pgellert @michael-redpanda @BenPope I dropped the commit that ben was commenting about w.r.t. using a `security_config` library approach. I think that's a good idea given how much sub-system validation code is re-used. I'll refactor that work and re-submit. Fixed the gellerts comment.
